keps/sig-network/3698-multi-network/README.md

```diff
@@ -0,0 +1,363 @@
+namespaces and network access by overriding the Cluster Default PodNetwork for
+a given Namespace. I do not want the tenants to change their manifests for that
+purpose. Those workloads should have the same level of Kubernetes functionality:
+Services, NetworkPolicies, access to Kubernetes API. I wish to support
+“profiles” for Namespace, where I can not only change Cluster Default PodNetwork,
+but define a set of PodNetworks assigned to given Namespace that Pods created in
+that Namespace are automatically attached to.
+
+<p align="center">
+  <img src="mn-story-6.png?raw=true" alt="multi-network story 6 namespace networks profiles"/>
+</p>
+
+#### Story #7
+As a “Power User” with admin privileges I wish to have the ability to modify my
+Pod network namespace without any restrictions. I am aware that by doing this I
+might break established contracts for the kubernetes features.
+
+#### Story #8
+As a Virtual Machine -based compute platform provider that I run on top of
+Kubernetes and Kubevirt I need the ability to add/remove Network Interfaces to
+existing VMs without re-creating the VM. This would not be applicable for the
+Primary PodNetwork.
+
+<p align="center">
+  <img src="mn-story-8.png?raw=true" alt="multi-network story 8 network interface hot-plug"/>
+</p>
+
+#### Story #9
+As an infrastructure administrator I wish to support IPv6 SLAAC in the
+Kubernetes Pods. SLAAC's nature is to dynamically assign IP addresses and change
+them over time. Those should be reflected in Pod status. This would not be
+applicable for the Primary PodNetwork.
+
+### Requirements
+
+Below requirements are divided into phases. Each phase will be in its own design
+section of this KEP that will be defined and implemented incrementally.
+
+#### Phase I (base API and reference in Pod)
+1. Multi-network should not change the behavior of existing clusters where no 
+multi-network constructs are configured/used.
+2. Multi-network needs to be represented explicitly as an API construct in
+Kubernetes as it fundamentally ties in with other APIs in the Kubernetes
+networking model, extending their functionality.
+   * We will introduce an PodNetwork API object that represents the existing
+     infrastructure’s networking.
+3. PodNetwork functions as a handle that decouples the implementation
+(i.e. infrastructure/network admins) from the consumer (cluster user). This
+abstraction follows the Kubernetes model of decoupling intent from implementation.
+4. PodNetwork shall not define any implementation specific parameters in its
+specification.
+5. PodNetwork shall provide cluster generic options only.
+6. PodNetwork can reference to implementation-specific parameters.
+7. PodNetwork is the object Application developer will reference in their
+workflows (e.g. Pod, Service).
+8. Cluster Default PodNetwork is the PodNetwork the cluster has been created
+with initially.
+9. Cluster Default PodNetwork cannot be removed, but can be replaced.
```

Comment:
These things are tricky, what happens with the pods? do they magically happen to be part of other network? ... this will cause disruption on workloads

Reply:
These are requirements, not how they are going to be implemented. Is the requirement in itself incorrect?

Reply:
I don't see it incorrect, but hard to implement it, pods are ephemeral , either run or die, replacing its network can mean a lot of things ...

Reply:
Oh, this requirement is about ability to replace default PodNetwork object in the cluster (which would probably need some sort of migration/pod restarts), not about hot-plug an interface to a Pod. Thou that part is requirement 24 :).

Reply:
where is this requirement coming from?
I don't expect networks or network attachments to be mutable

Reply:
This comes from the current ability to change CNI or just its configuration in a cluster. In cases where MN is ignored, that is not applicable, but in cluster where all is build on PodNetwork, we need ability to replace Default's PodNetwork. This was hinted by @danwinship.